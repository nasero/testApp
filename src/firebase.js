import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

const firebaseConfig = {
    apiKey: 'AIzaSyBQ1cYtMlmidcIoRf3FxEzrO87KW17G3wM',
    authDomain: 'testapp-f1ef6.firebaseapp.com',
    projectId: 'testapp-f1ef6',
    storageBucket: 'testapp-f1ef6.appspot.com',
    messagingSenderId: '103885774192',
    appId: '1:103885774192:web:56838029614ac54b1b83b5'
};

firebase.initializeApp(firebaseConfig);

// utils
const db = firebase.firestore();
const auth = firebase.auth();

export { db, auth };
