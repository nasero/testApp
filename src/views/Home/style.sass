.home
    &__title
        color: #000,
        margin: 0
    &__subtitle
        margin-top: 0.5em
    &__counters-group
        display: flex,
        justify-content: space-between,
        margin-top: 2.5em
    &__counter
        align-content: space-between
    &__button
        margin-top: 4em