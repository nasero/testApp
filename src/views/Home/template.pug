.view.home
    h2.home__title Welcome!
    p.home__subtitle The last time you accessed was
    .home__counters-group
        testApp-counter.home__counter(:value="lastVisitCount.days", label="days")
        testApp-counter.home__counter(:value="lastVisitCount.hours", label="hours")
        testApp-counter.home__counter(:value="lastVisitCount.mins", label="minutes")
        testApp-counter.home__counter(:value="lastVisitCount.secs", label="seconds")
    testApp-button.testApp-button--light.home__button(text="LOGOUT", @testApp-button-clicked="logout")
