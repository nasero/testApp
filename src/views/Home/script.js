/**
 * Vista Home
 */

import testAppCounter from '@/components/testApp-counter';
import { auth, db } from '../../firebase';
export default {
    name: 'Home',
    components: {
        testAppCounter
    },
    data () {
        return {
            lastLogin: 0,
            lastVisitCount: {
                days: '00',
                hours: '00',
                mins: '00',
                secs: '00'
            }
        };
    },
    methods: {
        /* istanbul ignore next */
        async getLastVisit () {
            try {
                const docRef = await db.collection('Users').doc(auth.currentUser.uid);
                const doc = await docRef.get();
                this.lastLogin = doc.data().lastVisit ? doc.data().lastVisit : new Date().getTime();
                this.calculateTimeSpend();
            } catch (error) {
                console.log(error);
            }
        },
        async updateLastVisit () {
            try {
                await db.collection('Users').doc(auth.currentUser.uid).set({
                    lastVisit: new Date().getTime()
                });
            } catch (error) {
                console.error('Error writing document: ', error);
            };
        },
        async logout () {
            await this.updateLastVisit();
            await auth.signOut();
            this.$router.push('login');
        },
        calculateTimeSpend () {
            const now = new Date().getTime();
            const timeSpend = now - this.lastLogin;
            const dateSpend = new Date(timeSpend);
            Object.assign(this.lastVisitCount,
                {
                    days: String(dateSpend.getDate() - 1).padStart(2, '0'),
                    hours: String(Math.max(dateSpend.getHours() - 1, 0)).padStart(2, '0'),
                    mins: String(dateSpend.getMinutes()).padStart(2, '0'),
                    secs: String(dateSpend.getSeconds()).padStart(2, '0')
                });
        }
    },
    mounted () {
        this.getLastVisit();
    }
};
