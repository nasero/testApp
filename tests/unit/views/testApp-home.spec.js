import { createLocalVue, shallowMount } from '@vue/test-utils';
import Home from '@/views/Home';
import TestAppButton from '@/components/testApp-button';
import * as admin from '@/firebase';

const data = jest.fn(() => ({ lastVisit: '12345678' }));
const set = jest.fn();
const get = jest.fn(() => ({ exists: true, data }));
const doc = jest.fn(() => ({ get, set }));
const collection = jest.spyOn(admin.db, 'collection').mockReturnValue({ doc });

admin.auth = {
    currentUser: {
        uid: '123'
    }
};

const localVue = createLocalVue();

localVue.component('testApp-button', TestAppButton);

describe('Home - Snapshot', () => {
    it('Snapshot', () => {
        const wrapper = shallowMount(Home, {
            localVue
        });
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe('Home - methods', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(Home, {
            localVue
        });
    });

    it('getLastVisit', () => {
        wrapper.vm.getLastVisit();

        expect(collection).toHaveBeenCalled();
    });

    it('updateLastVisit', () => {
        wrapper.vm.updateLastVisit();

        expect(collection).toHaveBeenCalled();
    });

    it('calculateTimeSpend', () => {
        const spy = jest.spyOn(global, 'String');
        wrapper.vm.lastLogin = 1611085228420;

        wrapper.vm.calculateTimeSpend();

        expect(spy).toHaveBeenCalledTimes(4);
    });
});
