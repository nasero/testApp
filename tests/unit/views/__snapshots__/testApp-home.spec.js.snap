// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home - Snapshot Snapshot 1`] = `
<div
  class="view home"
>
  <h2
    class="home__title"
  >
    Welcome!
  </h2>
  <p
    class="home__subtitle"
  >
    The last time you accessed was
  </p>
  <div
    class="home__counters-group"
  >
    <testapp-counter-stub
      class="home__counter"
      label="days"
      value="00"
    />
    <testapp-counter-stub
      class="home__counter"
      label="hours"
      value="00"
    />
    <testapp-counter-stub
      class="home__counter"
      label="minutes"
      value="00"
    />
    <testapp-counter-stub
      class="home__counter"
      label="seconds"
      value="00"
    />
  </div>
  <testapp-button-stub
    class="testApp-button--light home__button"
    text="LOGOUT"
    type="button"
  />
</div>
`;
